Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h

```diff
@@ -2619,41 +2620,73 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
         case Array::Contiguous:
         case Array::ArrayStorage:
         case Array::SlowPutArrayStorage:
-            if (node->arrayMode().isEffectfulOutOfBounds())
+            if (arrayMode.isEffectfulOutOfBounds())
                 clobberWorld();
             makeHeapTopForNode(node);
             break;
         case Array::Int8Array:
-            setNonCellTypeForNode(node, SpecInt32Only);
+            if (node->op() == GetByVal && arrayMode.isOutOfBounds())
+                setNonCellTypeForNode(node, SpecInt32Only | SpecOther);
+            else
+                setNonCellTypeForNode(node, SpecInt32Only);
             break;
         case Array::Int16Array:
-            setNonCellTypeForNode(node, SpecInt32Only);
+            if (node->op() == GetByVal && arrayMode.isOutOfBounds())
+                setNonCellTypeForNode(node, SpecInt32Only | SpecOther);
+            else
+                setNonCellTypeForNode(node, SpecInt32Only);
             break;
         case Array::Int32Array:
-            setNonCellTypeForNode(node, SpecInt32Only);
+            if (node->op() == GetByVal && arrayMode.isOutOfBounds())
+                setNonCellTypeForNode(node, SpecInt32Only | SpecOther);
+            else
+                setNonCellTypeForNode(node, SpecInt32Only);
             break;
         case Array::Uint8Array:
-            setNonCellTypeForNode(node, SpecInt32Only);
+            if (node->op() == GetByVal && arrayMode.isOutOfBounds())
+                setNonCellTypeForNode(node, SpecInt32Only | SpecOther);
+            else
+                setNonCellTypeForNode(node, SpecInt32Only);
             break;
         case Array::Uint8ClampedArray:
-            setNonCellTypeForNode(node, SpecInt32Only);
+            if (node->op() == GetByVal && arrayMode.isOutOfBounds())
+                setNonCellTypeForNode(node, SpecInt32Only | SpecOther);
+            else
+                setNonCellTypeForNode(node, SpecInt32Only);
```

Comment:
This is specifically done to handle each case carefully and independently because they can have different optimizations like Int52. And for example, we are not listing Uint32Array here since it has special cases.